### DISCUSSION TITLE
Include inputPath in environment options passed to markdown

### DISCUSSION THREAD

**Discussion by GeorgeNance** [11ty/eleventy#4072] [Enhancement Queue]
**Is your feature request related to a problem? Please describe.**
Yes, It would be nice to have context to which file is being rendered when I am writing custom plugins for my markdown. 

I am trying to process images that are in the same folder as a markdown file.

**Describe the solution you'd like**

Simply passing the inputPath to `mdlib.render` in Markdown.js

**Describe alternatives you've considered**
There is no easy alternate at this point. 

**Additional context**
N/A

**Comment by zachleat** [comment]
Hmm—does `page` solve this? `page.inputPath`
https://www.11ty.dev/docs/data-eleventy-supplied/#page-variable-contents

**Comment by GeorgeNance** [comment]
I don’t believe we are able to access page variables from plugins that are called when rendering . I may be wrong .

**Comment by zachleat** [comment]
> This repository is now using [lodash style issue management](https://twitter.com/samselikoff/status/991395669016436736) for enhancements. This means enhancement issues will now be closed instead of leaving them open.

> View [the enhancement backlog here](https://github.com/11ty/eleventy/issues?utf8=%E2%9C%93&q=label%3Aneeds-votes+sort%3Areactions-%2B1-desc+). Don’t forget to upvote the top comment with 👍!

**Comment by manunamz** [comment]
For others who also want to access `inputPath` in a markdown-it plugin:

This is already possible, if a bit loopy. See details below.

- 11ty stuffs its `data` into the markdown-it library's render function [here](https://github.com/11ty/eleventy/blob/master/src/Engines/Markdown.js#L85).
- Markdown-it's `render` function receives it [here](https://github.com/markdown-it/markdown-it/blob/df4607f1d4d4be7fdc32e71c04109aea8cc373fa/lib/index.js#L541), with [this](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/cd29f1c657713344b002f8a87de631cb3e60989a/types/markdown-it/lib/index.d.ts#L298) type declaration.
- You may then access `env.page.inputPath` (among other 11ty supplied data) inside of your render rules, which might look something like:

```js
function customPlugin (md, opts) {
  md.renderer.rules.custom_render_rule = function (tokens, index, opts, env) {
    // do some things...
    return `now you can access inputPath! ${env.page.inputPath}`
  }
}
```

### However...

...you should keep in mind that by accessing `env.page` you are tying your plugin impelementation to 11ty. To make this more universally accessible, you might want to add a configuration such as:
```js
const customPluginDefaults = {
  getAbsPathFromEnv: (env) => { return env.page.inputPath }, // in the case of 11ty
}
```

...which would change the render rule to look like:

```js
function customPlugin (md, opts) {
  const fullOptions = { ...customPluginDefaults, ...opts }; // this line merges default options with custom options

  md.renderer.rules.custom_render_rule = function (tokens, index, opts, env) {
    // do some things...
    return `now you can access inputPath! ${fullOptions.getAbsPathFromEnv(env)}`
  }
}
```

EDIT:

A live example plugin can be viewed at: [`markdown-it-wikirefs`](https://github.com/wikibonsai/markdown-it-wikirefs/); see its [`resolveHtmlHref`](https://github.com/wikibonsai/markdown-it-wikirefs/tree/main#resolvehtmlhref-env-any-fname-string--string--undefined) option.

See it in action in [`eleventy-wikibonsai`](https://github.com/wikibonsai/eleventy-wikibonsai/blob/main/wikibonsai/markdown.js#L15), which accesses the `inputPath` in its configuration.

**Comment by nhoizey** [comment]
And now there's also `markdown-it-image-size` thanks to @boyum:
- https://github.com/boyum/markdown-it-image-size/issues/404#issuecomment-1483759243
